extensions/wikia/AdEngine/js/context/adContext.js

```diff
@@ -42,11 +42,67 @@ define('ext.wikia.adEngine.adContext', [
 		return context.targeting.pageType === pageType;
 	}
 
+	function isPageFairDetectionEnabled() {
+		var isSupportedGeo = geo.isProperGeo(instantGlobals.wgAdDriverPageFairDetectionCountries);
+		return isUrlParamSet('pagefairdetection') || (isSupportedGeo && sampler.sample('pageFairDetection', 1, 10));
+	}
+
+	function isGSCEnabled() {
+		return abTest.getGroup('PROJECT_43') === 'GROUP_5' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverGoogleConsumerSurveysCountries);
+	}
+
+	function isSourcePointDetectionDesktopEnabled(context) {
+		return context.opts.sourcePointDetectionUrl && (context.targeting.skin === 'oasis' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionCountries));
+	}
+
+	function isSourcePointDetectionMobile(context) {
+		return context.opts.sourcePointDetectionUrl && (context.targeting.skin === 'mercury' &&
+			geo.isProperGeo(instantGlobals.wgAdDriverSourcePointDetectionMobileCountries));
+	}
+
+	function updateRecoveryServicesAdContext(context, noExternals) {
+		var taboolaConfig = instantGlobals.wgAdDriverTaboolaConfig || {},
+			isRecoveryServiceEnabled = false;
+
+		// PageFair recovery
+		context.opts.pageFairRecovery = !noExternals && !isRecoveryServiceEnabled &&
+			context.opts.pageFairRecovery && geo.isProperGeo(instantGlobals.wgAdDriverPageFairRecoveryCountries);
+		isRecoveryServiceEnabled = context.opts.pageFairRecovery;
+
+		// PageFair detection
+		context.opts.pageFairDetection = !noExternals && isPageFairDetectionEnabled(context);
+
+		// SourcePoint recovery
+		context.opts.sourcePointRecovery = !noExternals && !isRecoveryServiceEnabled &&
+			context.opts.sourcePointRecovery && geo.isProperGeo(instantGlobals.wgAdDriverSourcePointRecoveryCountries);
+		isRecoveryServiceEnabled = isRecoveryServiceEnabled || context.opts.sourcePointRecovery;
+
+		// SourcePoint MMS
+		context.opts.sourcePointMMS = !noExternals && !isRecoveryServiceEnabled && context.opts.sourcePointMMS;
+		isRecoveryServiceEnabled = isRecoveryServiceEnabled || context.opts.sourcePointMMS;
+
+		context.opts.sourcePointBootstrap = context.opts.sourcePointMMS || context.opts.sourcePointRecovery;
+
+		// SourcePoint detection integration
+		context.opts.sourcePointDetection = !noExternals && isSourcePointDetectionDesktopEnabled(context);
+		context.opts.sourcePointDetectionMobile = !noExternals && isSourcePointDetectionMobile(context);
+
+		// Taboola
+		context.opts.loadTaboolaLibrary = !noExternals && !isRecoveryServiceEnabled &&
+			shouldLoadTaboolaOnBlockingTraffic(taboolaConfig);
+		isRecoveryServiceEnabled = isRecoveryServiceEnabled || context.opts.loadTaboolaLibrary;
+
+		// Google Consumer Surveys
+		context.opts.googleConsumerSurveys = context.opts.sourcePointDetection &&
+			!isRecoveryServiceEnabled && context.opts.showAds && isGSCEnabled();
+	}
+
 	function setContext(newContext) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Function has a complexity of 33 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)